File: lib/server/models/Review.js
const mongoose = require('mongoose');

const reviewSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },

  username: {
    type: String,
    required: true
  },

  userAvatar: {
    type: String,
    default: ''
  },

  itemId: {
    type: mongoose.Schema.Types.ObjectId,
    required: true,
    index: true
  },

  itemType: {
    type: String,
    enum: ['instrument', 'stage', 'studio'],
    required: true
  },

  rating: {
    type: Number,
    required: true,
    min: 1,
    max: 5
  },

  comment: {
    type: String,
    trim: true,
    maxlength: 1000
  },

  createdAt: {
    type: Date,
    default: Date.now
  }
});

// One review per user per item
reviewSchema.index({ userId: 1, itemId: 1, itemType: 1 }, { unique: true });
reviewSchema.index({ itemId: 1, itemType: 1, createdAt: -1 });

// After save — recalculate rating on the item
async function recalcRating(doc) {
  const reviews = await mongoose.model('Review').find({
    itemId: doc.itemId,
    itemType: doc.itemType
  });

  if (reviews.length === 0) return;

  const avg = reviews.reduce((sum, r) => sum + r.rating, 0) / reviews.length;
  const rounded = Math.round(avg * 10) / 10;

  let ItemModel;
  switch (doc.itemType) {
    case 'instrument': ItemModel = mongoose.model('Instrument'); break;
    case 'stage':      ItemModel = mongoose.model('Stage');      break;
    case 'studio':     ItemModel = mongoose.model('Studio');     break;
  }

  if (ItemModel) {
    await ItemModel.findByIdAndUpdate(doc.itemId, {
      rating: rounded,
      reviewsCount: reviews.length
    });
  }
}

reviewSchema.post('save', recalcRating);
reviewSchema.post('findOneAndDelete', recalcRating);

module.exports = mongoose.model('Review', reviewSchema);

File: lib/models/review.dart
class Review {
  final String id;
  final String userId;
  final String username;
  final String userAvatar;
  final String itemId;
  final String itemType;
  final int rating;
  final String comment;
  final DateTime createdAt;

  Review({
    required this.id,
    required this.userId,
    required this.username,
    this.userAvatar = '',
    required this.itemId,
    required this.itemType,
    required this.rating,
    this.comment = '',
    required this.createdAt,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['_id'] ?? json['id'] ?? '',
      userId: json['userId']?.toString() ?? '',
      username: json['username'] ?? 'Қолданушы',
      userAvatar: json['userAvatar'] ?? '',
      itemId: json['itemId']?.toString() ?? '',
      itemType: json['itemType'] ?? '',
      rating: (json['rating'] ?? 1).toInt(),
      comment: json['comment'] ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  String get initials {
    if (username.isEmpty) return '?';
    final parts = username.split(' ');
    if (parts.length >= 2) {
      return '${parts[0][0]}${parts[1][0]}'.toUpperCase();
    }
    return username.substring(0, username.length > 1 ? 2 : 1).toUpperCase();
  }
}
